cmd/frontend/frontend.go

```diff
@@ -0,0 +1,172 @@
+
+	SubscriptionPath  = "/subscriptions/{subscriptionId}"
+	ResourceGroupPath = SubscriptionPath + "/resourcegroups/{resourceGroupName}"
+	ResourceTypePath  = ResourceGroupPath + "/" + ResourceProviderNamespace + "/{resourceType}"
+	ResourceNamePath  = ResourceTypePath + "/{resourceName}"
+)
+
+type PathValues struct {
+	SubscriptionID    string
+	ResourceGroupName string
+	ResourceType      string
+	ResourceName      string
+}
+
+func NewPathValues(request *http.Request) *PathValues {
+	return &PathValues{
+		SubscriptionID:    request.PathValue("subscriptionId"),
+		ResourceGroupName: request.PathValue("resourceGroupName"),
+		ResourceType:      request.PathValue("resourceType"),
+		ResourceName:      request.PathValue("resourceName"),
+	}
+}
+
+type Frontend struct {
+	logger   *slog.Logger
+	listener net.Listener
+	server   http.Server
+	ready    atomic.Value
+	done     chan struct{}
+}
+
+func NewFrontend(logger *slog.Logger, listener net.Listener) *Frontend {
+	f := &Frontend{
+		logger:   logger,
+		listener: listener,
+		server: http.Server{
+			ErrorLog: slog.NewLogLogger(logger.Handler(), slog.LevelError),
+		},
+		done: make(chan struct{}),
+	}
+
+	mux := NewMiddlewareMux(
+		MiddlewarePanic,
+		MiddlewareBody,
+		MiddlewareLowercase,
+		MiddlewareSystemData)
+
+	// Unauthenticated routes
+	mux.HandleFunc("/", f.NotFound)
+	mux.HandleFunc("GET /healthz/ready", f.HealthzReady)
+
+	// Authenticated routes
+	postMuxMiddleware := NewMiddleware(MiddlewareValidateAPIVersion)
+	mux.Handle(
+		"GET "+ResourceTypePath,
+		postMuxMiddleware.HandlerFunc(f.ArmResourceListByParent))
+	mux.Handle(
+		"GET "+ResourceNamePath,
+		postMuxMiddleware.HandlerFunc(f.ArmResourceRead))
+	mux.Handle(
+		"PUT "+ResourceNamePath,
+		postMuxMiddleware.HandlerFunc(f.ArmResourceCreateOrUpdate))
+	mux.Handle(
+		"PATCH "+ResourceNamePath,
+		postMuxMiddleware.HandlerFunc(f.ArmResourcePatch))
+	mux.Handle(
+		"DELETE "+ResourceNamePath,
+		postMuxMiddleware.HandlerFunc(f.ArmResourceDelete))
+	mux.Handle(
+		"POST "+ResourceNamePath,
+		postMuxMiddleware.HandlerFunc(f.ArmResourceAction))
+	f.server.Handler = mux
+
+	return f
+}
+
+func (f *Frontend) Run(ctx context.Context, stop <-chan struct{}) {
+	if stop != nil {
+		go func() {
+			<-stop
+			f.ready.Store(false)
+			f.server.Shutdown(ctx)
+		}()
+	}
+
+	f.logger.Info(fmt.Sprintf("listening on %s", f.listener.Addr().String()))
+
+	f.ready.Store(true)
+
+	err := f.server.Serve(f.listener)
+	if err != http.ErrServerClosed {
+		f.logger.Error(err.Error())
+		os.Exit(1)
+	}
+
+	close(f.done)
+}
+
+func (f *Frontend) Join() {
+	<-f.done
+}
+
+func (f *Frontend) CheckReady() bool {
+	return f.ready.Load().(bool)
+}
+
+func (f *Frontend) NotFound(writer http.ResponseWriter, request *http.Request) {
+	arm.WriteError(
+		writer, http.StatusNotFound,
+		arm.CloudErrorCodeNotFound, "",
+		"The requested path could not be found.")
+}
+
+func (f *Frontend) HealthzReady(writer http.ResponseWriter, request *http.Request) {
```

Comment:
I think for `healthz` this isn't an API contract we adhere to, so just returning a 200 for ready vs. a 500 for not ready would be fine.  No need to format it according to ARM response guidelines.

Reply:
Ok.  I was blindly copying from ARO-RP there.